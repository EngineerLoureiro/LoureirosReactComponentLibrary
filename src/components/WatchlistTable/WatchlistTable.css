.card {
  display: grid;
  border-radius: 2rem;
  box-shadow: 0px 10px 15px -3px rgba(0, 0, 0, 0.1);
  max-width: 100%;
  width: fit-content;
  padding: 1rem;
}

.table {
  color: #000;
  border-collapse: collapse;
  grid-template-columns: repeat(4, 1fr);
}

.table-body td {
  padding: 0.5rem 0.5rem 0 0.5rem;
}

.table-item:not(:last-child) {
  border-bottom: 1px solid #ccc;
}
