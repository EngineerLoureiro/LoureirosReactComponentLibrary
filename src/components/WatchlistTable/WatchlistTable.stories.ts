// Replace your-framework with the framework you are using, e.g. react-vite, nextjs, nextjs-vite, etc.
import type { Meta, StoryObj } from "@storybook/react-vite";

import { WatchlistTable, WatchlistTableItemProps } from "./WatchlistTable";

import { fn } from "storybook/test";

//ðŸ‘‡ This default export determines where your story goes in the story list
const meta = {
  title: "Loureiro'sComponentLibrary/WatchlistTable",
  component: WatchlistTable,
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/configure/story-layout
    layout: "centered",
  },
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/writing-docs/autodocs
  tags: ["autodocs"],
  args: {},
} satisfies Meta<typeof WatchlistTable>;

export default meta;
type Story = StoryObj<typeof meta>;

const btcQuoteItems: WatchlistTableItemProps[] = [
  {
    symbol: "SOLBTC",
    lastPrice: "0.00310200",
    lastPriceChange: "+0.00002100",
    lastPriceChangePercentage: "+0.68%",
  },
  {
    symbol: "BNBBTC",
    lastPrice: "0.00918000",
    lastPriceChange: "-0.00006000",
    lastPriceChangePercentage: "-0.65%",
  },
  {
    symbol: "XRPBTC",
    lastPrice: "0.00000980",
    lastPriceChange: "+0.00000010",
    lastPriceChangePercentage: "+1.03%",
  },
  {
    symbol: "DOGEBTC",
    lastPrice: "0.00000200",
    lastPriceChange: "-0.00000003",
    lastPriceChangePercentage: "-1.48%",
  },
  {
    symbol: "ADABTC",
    lastPrice: "0.00000710",
    lastPriceChange: "+0.00000008",
    lastPriceChangePercentage: "+1.14%",
  },
  {
    symbol: "AVAXBTC",
    lastPrice: "0.00064000",
    lastPriceChange: "-0.00000850",
    lastPriceChangePercentage: "-1.31%",
  },
  {
    symbol: "LINKBTC",
    lastPrice: "0.00025600",
    lastPriceChange: "+0.00000340",
    lastPriceChangePercentage: "+1.34%",
  },
  {
    symbol: "MATICBTC",
    lastPrice: "0.00001220",
    lastPriceChange: "-0.00000015",
    lastPriceChangePercentage: "-1.22%",
  },
  {
    symbol: "DOTBTC",
    lastPrice: "0.00008600",
    lastPriceChange: "+0.00000140",
    lastPriceChangePercentage: "+1.65%",
  },
  {
    symbol: "ATOMBTC",
    lastPrice: "0.00012000",
    lastPriceChange: "-0.00000120",
    lastPriceChangePercentage: "-0.99%",
  },
  {
    symbol: "LTCBTC",
    lastPrice: "0.00110000",
    lastPriceChange: "+0.00000690",
    lastPriceChangePercentage: "+0.63%",
  },
  {
    symbol: "NEARBTC",
    lastPrice: "0.00011000",
    lastPriceChange: "+0.00000120",
    lastPriceChangePercentage: "+1.10%",
  },
];

export const DefaultWatchlist: Story = {
  args: {
    items: btcQuoteItems,
  },
};
