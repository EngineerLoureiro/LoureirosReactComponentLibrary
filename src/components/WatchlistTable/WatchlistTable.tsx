import "./WatchlistTable.css";

export interface WatchlistTableItemProps {
  symbol: string;
  lastPrice: string;
  lastPriceChange: string;
  lastPriceChangePercentage: string;
}

export interface WatchlistTableProps {
  items: WatchlistTableItemProps[];
}

export function WatchlistTable({ items }: WatchlistTableProps) {
  return (
    <div className="card">
      <table className="table">
        <thead>
          <tr>
            <th>{"Symbol"}</th>
            <th>{"Price"}</th>
            <th>{"Chg"}</th>
            <th>{"Chg%"}</th>
          </tr>
        </thead>
        <tbody className="table-body">
          {items.map((row) => {
            return (
              <tr className="table-item">
                <td>{row.symbol}</td>
                <td>{row.lastPrice}</td>
                <td>{row.lastPriceChange}</td>
                <td>{row.lastPriceChangePercentage}</td>
              </tr>
            );
          })}
        </tbody>
      </table>
    </div>
  );
}
