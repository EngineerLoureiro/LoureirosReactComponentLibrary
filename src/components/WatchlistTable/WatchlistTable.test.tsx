// WatchlistTable.test.tsx
import { render, screen, within } from "@testing-library/react";
import { WatchlistTable, WatchlistTableItemProps } from "./WatchlistTable";

const initialItems: WatchlistTableItemProps[] = [
  {
    symbol: "SOLBTC",
    lastPrice: "0.00310200",
    lastPriceChange: "+0.00002100",
    lastPriceChangePercentage: "+0.68%",
  },
  {
    symbol: "BNBBTC",
    lastPrice: "0.00918000",
    lastPriceChange: "-0.00006000",
    lastPriceChangePercentage: "-0.65%",
  },
  {
    symbol: "LINKETH",
    lastPrice: "0.005000",
    lastPriceChange: "+0.000067",
    lastPriceChangePercentage: "+1.35%",
  },
];

const updatedItems: WatchlistTableItemProps[] = [
  {
    symbol: "SOLBTC",
    lastPrice: "0.00300200",
    lastPriceChange: "-0.00010000",
    lastPriceChangePercentage: "-3.22%",
  },
  {
    symbol: "BNBBTC",
    lastPrice: "0.00920000",
    lastPriceChange: "+0.00002000",
    lastPriceChangePercentage: "+0.22%",
  },
  {
    symbol: "LINKETH",
    lastPrice: "0.004950",
    lastPriceChange: "-0.000050",
    lastPriceChangePercentage: "-1.00%",
  },
];

describe("WatchlistTable", () => {
  test("renders a semantic table with headers", () => {
    render(<WatchlistTable items={initialItems} />);

    // table exists
    const table = screen.getByRole("table");
    expect(table).toBeInTheDocument();

    // headers
    const headers = screen.getAllByRole("columnheader");
    expect(headers).toHaveLength(4);
    expect(headers.map((h) => h.textContent)).toEqual([
      "Symbol",
      "Price",
      "Chg",
      "Chg%",
    ]);

    // header row + 3 body rows
    const allRows = screen.getAllByRole("row");
    expect(allRows.length).toBe(1 + initialItems.length);
  });

  test("renders all rows and expected cell content", () => {
    render(<WatchlistTable items={initialItems} />);

    const table = screen.getByRole("table");
    const tbody = table.querySelector("tbody")!; // tbody
    const bodyRows = within(tbody).getAllByRole("row");
    expect(bodyRows).toHaveLength(initialItems.length);

    // check a couple of cells
    expect(screen.getByText("SOLBTC")).toBeInTheDocument();
    expect(screen.getByText("0.00918000")).toBeInTheDocument(); // BNBBTC lastPrice
    expect(screen.getByText("+1.35%")).toBeInTheDocument(); // LINKETH change%
  });

  test("updates UI when items prop changes (rerender)", () => {
    const { rerender } = render(<WatchlistTable items={initialItems} />);

    // precondition
    expect(screen.getByText("0.00310200")).toBeInTheDocument();

    // update props
    rerender(<WatchlistTable items={updatedItems} />);

    // old value gone, new value present
    expect(screen.queryByText("0.00310200")).not.toBeInTheDocument();
    expect(screen.getByText("0.00300200")).toBeInTheDocument();

    // also verify a changed percentage is reflected
    expect(screen.getByText("-3.22%")).toBeInTheDocument();
  });

  test("matches snapshot (with data)", () => {
    const { asFragment } = render(<WatchlistTable items={initialItems} />);
    expect(asFragment()).toMatchSnapshot();
  });

  test("matches snapshot (empty state)", () => {
    const { asFragment } = render(<WatchlistTable items={[]} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
