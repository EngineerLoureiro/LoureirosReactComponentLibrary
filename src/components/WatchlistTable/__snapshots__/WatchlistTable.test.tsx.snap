// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WatchlistTable matches snapshot (empty state) 1`] = `
<DocumentFragment>
  <div
    class="card"
  >
    <table
      class="table"
    >
      <thead>
        <tr>
          <th>
            Symbol
          </th>
          <th>
            Price
          </th>
          <th>
            Chg
          </th>
          <th>
            Chg%
          </th>
        </tr>
      </thead>
      <tbody
        class="table-body"
      />
    </table>
  </div>
</DocumentFragment>
`;

exports[`WatchlistTable matches snapshot (with data) 1`] = `
<DocumentFragment>
  <div
    class="card"
  >
    <table
      class="table"
    >
      <thead>
        <tr>
          <th>
            Symbol
          </th>
          <th>
            Price
          </th>
          <th>
            Chg
          </th>
          <th>
            Chg%
          </th>
        </tr>
      </thead>
      <tbody
        class="table-body"
      >
        <tr
          class="table-item"
        >
          <td>
            SOLBTC
          </td>
          <td>
            0.00310200
          </td>
          <td>
            +0.00002100
          </td>
          <td>
            +0.68%
          </td>
        </tr>
        <tr
          class="table-item"
        >
          <td>
            BNBBTC
          </td>
          <td>
            0.00918000
          </td>
          <td>
            -0.00006000
          </td>
          <td>
            -0.65%
          </td>
        </tr>
        <tr
          class="table-item"
        >
          <td>
            LINKETH
          </td>
          <td>
            0.005000
          </td>
          <td>
            +0.000067
          </td>
          <td>
            +1.35%
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
